mdtools/__init__.py

```diff
@@ -63,3 +64,26 @@
 import mdtools.select
 import mdtools.statistics as stats
 import mdtools.structure as strc
+
+__all__ = [
+    "__version__",
+    "__title__",
+    "__author__",
+    "__copyright__",
+    "__copyright_notice__",
+    "box",
+    "check",
+    "dtrj",
+    "dyn",
+    "fh",
+    "func",
+    "msm",
```

Comment:
Now, this is actually interesting. This should have been caught by [F822](https://www.flake8rules.com/rules/F822.html) which should be active in the workflow (via F82 wildcard).

Reply:
This needs some rework. The exported symbols are not quite right and I don't yet fully understand the visibility of imports in python modules.

Reply:
I do not really get what you mean with "exported symbols".

Regarding imports in Python, that's indeed a bit tricky. Same with the `__all__` dunder ("magic methods/variables in Python). In fact, the `__all__` variable only tells you what submodules/classes/functions/variables get imported when you do `from mdtools import *`. But these "asterisk-imports" are highly discouraged anyhow. The `__all__` variable does not prevent you from importing submodules that are not defined in the `__all__` variable. So if you would remove `"func"` from `__all__`, I could still do `from mdtools import func`. See also the first four bullet points in [this post](http://xion.io/post/code/python-all-wild-imports.html). So, you do not need to define an `__all__` variable, but many modules do so.

Regarding the imports in the `__init__.py`: If I would not have imported the submodules there, you would have to import them yourself explicitly. This means in order to use the `func` submodule you would have to do

    from mdtools import func
    func.some_function_defined_in_func(args)

Only when the `func` submodule is imported in the `__init__.py`, you can do

    import mdtools
    mdtools.func.some_function_defined_in_func(args)

Reply:
> This should have been caught by F822

Is [F822](https://www.flake8rules.com/rules/F822.html) selected in the workflow? I do not see it there, or is it covered by F82?

    flake8 . --count --select=E1,E2,E3,E4,E9,F63,F4,F7,F82,W1,W2,W3 --show-source --statistics

Reply:
> So, you do not need to define an __all__ variable, but many modules do so.

Thanks for the detailed explanation. I'm missing the python specific nomenclature. With 'symbols' I meant those wildcard imports with `import *`. The introduction of the `__all__` dunder is for two reasons: First, it seems good practice (as you said, many modules do so). Second, it alleviates the "unused imports" warning in `__init__.py`.
Unfortunately, flake8 does not seem to recognize the usage in all cases. E.g. `import mdtools.select` has to be `from mdtools import select` in order to suppress the linter warning which should be functionally the same. This is the reason why I added `# flake8: noqa` but in hindsight it would make more sense to change the imports just to satisfy flake8.

> I do not see it there, or is it covered by F82?

It's covered by F82 which expands to all F codes that start with 82.
Unfortunately the F822 lint seems to be a separate issue. I can not get this to work even with [this simplest example](https://www.flake8rules.com/rules/F822.html). So this might be a defect of flake8...

Reply:
If `from mdtools import select` fixes the linter warning, you can change it. It's the same as `import mdtools.select`.

Reply:
Changed the import format as suggested. So this should not emit any warnings about unused imports.

This F822 thing seems to be jet another puzzling python quirk and I have given up on this. `pyflakes` [explicitly allows (see tests)](https://github.com/PyCQA/pyflakes/blob/4cf21896576b734fd901b0dbf6efb0a15162b4af/pyflakes/test/test_imports.py#L1140-L1145) unused names in `__all__` and I could not quickly find any reason why.

There seems to be some arguments about whether `__version__` and the like should be included in `__all__` (see pyflakes PR 645 [I don't want to links explicitly to avoid the auto linker]). Including everything should be fine for our uses since wildcard imports are discouraged anyway.

Reply:
Ok, then just leave it as it is.